File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        // int ans=0;
        // for(int i=0;i<s.length();i++){
        //     Set<Character> set=new HashSet<>();
        //     for(int j=i;j<s.length();j++){
        //         if(set.contains(s.charAt(j)))
        //            break;
        //         set.add(s.charAt(j));
        //     }
        //     ans=Math.max(ans,set.size());
        //     if(s.length()-i-1==ans) return ans;
        // }
        Map<Character,Integer> map=new HashMap<>();
        int ans=0,si=0;
        for(int i=0;i<s.length();i++){
            if(map.containsKey(s.charAt(i))){
               si=Math.max(map.get(s.charAt(i))+1,si);
            }
            System.out.println(si+" "+i);
            map.put(s.charAt(i),i);
            ans=Math.max(ans,i-si+1);
        }

        return ans;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        index = {}
        l = 0
        ans = 0

        for i in range(len(s)):
            c = s[i]
            if c in index:
                l = max(l, index[c] + 1)
            
            ans = max(ans, i - l + 1)
            index[c] = i
            print(i, ans)
        
        return ans

        